From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Mon, 25 Aug 2025 09:38:03 +0000
Subject: [PATCH 103/243] NVIDIA: SAUCE: tegra_hv: clear out memory before
 release

- clear out mempool memory while unreserving mempool

Bug 5433066

Change-Id: I032157dba3757df760ba8977f9ad720df4c8429c
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3449493
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
Reviewed-by: Sandeep Trasi <strasi@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
(cherry picked from commit 9670b0e58fa4d81733edde9773b91904a9508926)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index 9b1be4d1db2b5..5404738330b87 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -1063,15 +1063,45 @@ EXPORT_SYMBOL(tegra_hv_mempool_reserve);
 
 int tegra_hv_mempool_unreserve(struct tegra_hv_ivm_cookie *ivmk)
 {
-	int reserved;
-	struct hv_mempool *mempool = container_of(ivmk, struct hv_mempool,
-			ivmk);
+	int reserved = 0;
+	void *vaddr = NULL;
+	struct hv_mempool *mempool;
+
+	if (ivmk == NULL) {
+		ERR("error: ivmk is NULL\n");
+		return -EINVAL;
+	}
+
+	mempool = container_of(ivmk, struct hv_mempool, ivmk);
+	/* Validate mempool structure */
+	if (mempool == NULL || mempool->mpd == NULL) {
+		ERR("error: invalid mempool structure\n");
+		return -EINVAL;
+	}
+
+	/* Validate cookie consistency */
+	if (ivmk->size != mempool->mpd->size || mempool->mpd->pa != ivmk->ipa) {
+		ERR("error: invalid ivmk cookie\n");
+		return -EINVAL;
+	}
+
+	/* Get VA for the IPA */
+	vaddr = memremap(ivmk->ipa, ivmk->size, MEMREMAP_WB);
+	if (vaddr == NULL) {
+		ERR("error: failed to map memory\n");
+		return -ENOMEM;
+	}
 
 	mutex_lock(&mempool->lock);
+	/* clear mempool memory before unreserving mempool */
+	memset(vaddr, 0, ivmk->size);
 	reserved = mempool->reserved;
 	mempool->reserved = 0;
 	mutex_unlock(&mempool->lock);
 
+	/* Clean up mapped memory */
+	memunmap(vaddr);
+
 	return reserved ? 0 : -EINVAL;
 }
 EXPORT_SYMBOL(tegra_hv_mempool_unreserve);
-- 
2.43.0

